tests/FINDOLOGIC/Tests/DataElementsTest.php

```diff
@@ -0,0 +1,121 @@
+            'Keyword with value' => array('value', Keyword::class, false),
+            'Ordernumber with empty value' => array('', Ordernumber::class, true),
+            'Ordernumber with value' => array('value', Ordernumber::class, false)
+        );
+    }
+
+    /**
+     * @dataProvider multiValueItemProvider
+     */
+    public function testAddingEmptyValuesToMultiValueItemCausesException($value = '', $elementtype, $shouldCauseException = true)
+    {
+        try {
+            $element = new $elementtype($value);
+            if ($shouldCauseException) {
+                $this->fail('Adding empty values should cause exception!');
+            }
+        } catch (\Exception $exception) {
+            $this->assertEquals(EmptyValueNotAllowedException::class, get_class($exception));
+        }
+    }
+
+    public function simpleValueItemProvider()
+    {
+        return array(
+            'Bonus with empty value' => array('', Bonus::class, true),
+            'Bonus with value' => array('value', Bonus::class, false),
+            'Description with empty value' => array('', Description::class, true),
+            'Description with value' => array('value', Description::class, false),
+            'Name with empty value' => array('', Name::class, true),
+            'Name with value' => array('value', Name::class, false),
+            'Price with empty value' => array('', Price::class, true),
+            'Price with value' => array('value', Price::class, false),
+            'SalesFrequency with empty value' => array('', SalesFrequency::class, true),
+            'SalesFrequency with value' => array('value', SalesFrequency::class, false),
+            'Sort with empty value' => array('', Sort::class, true),
+            'Sort with value' => array('value', Sort::class, false),
+            'Summary with empty value' => array('', Summary::class, true),
+            'Summary with value' => array('value', Summary::class, false),
+            'Url with empty value' => array('', Url::class, true),
+            'Url with value' => array('value', Url::class, false)
+        );
+    }
+
+    /**
+     * @dataProvider simpleValueItemProvider
+     */
+    public function testAddingEmptyValuesToSimpleItemsCausesException($value = '', $elementType, $shouldCauseException = true)
+    {
+        try {
+            $element = new $elementType();
+            $element->setValue($value);
+            if ($shouldCauseException) {
+                $this->fail('Adding empty values should cause exception!');
+            }
+        } catch (\Exception $exception) {
+            $this->assertEquals(EmptyValueNotAllowedException::class, get_class($exception));
+        }
+    }
+
+    /**
+     * @expectedException \BadMethodCallException
+     */
+    public function testCallingSetValueMethodOfDateAddedClassCausesException()
+    {
+        $element = new DateAdded();
+        $element->setValue("");
+    }
+
+    public function emptyValueProvider()
```

Comment:
Add a doc block describing the structure that each data set has.

Reply:
Done

Reply:
What I meant was listing a description of which values is being provided. In this case, something like element value, element class, and whether that value for this particular class should cause an exception. In other words: What's the structure of this mysterious array?

Reply:
Done